plugins/GSdx/GSTextureCache.cpp

```diff
@@ -2070,16 +2074,34 @@ void GSTextureCache::SourceMap::RemoveAt(Source* s)
 				s->m_texture ? s->m_texture->GetID() : 0,
 				s->m_TEX0.TBP0);
 
-	// Source (except render target) is duplicated for each page they use.
-	for(size_t start = s->m_TEX0.TBP0 >> 5, end = s->m_target ? start : countof(m_map) - 1; start <= end; start++)
+	if (s->m_target)
 	{
-		list<Source*>& m = m_map[start];
+		size_t page = s->m_TEX0.TBP0 >> 5;
+		m_map[page].erase(s->m_erase_it[page]);
 
-		for(list<Source*>::iterator i = m.begin(); i != m.end(); )
+	}
+	else
+	{
+		// Mirror of GetPagesCoverage
+		for(size_t i = 0; i < countof(m_pages); i++)
 		{
-			list<Source*>::iterator j = i++;
+			if(uint32 p = s->m_pages_ptr[i])
+			{
+				list<Source*>* m = &m_map[i << 5];
+				auto* e = &s->m_erase_it[i << 5];
+
+				unsigned long j;
 
-			if(*j == s) {m.erase(j); break;}
+				while(_BitScanForward(&j, p))
+				{
+					// FIXME: this statement could be optimized to a single ASM instruction (instead of 4)
+					// Either BTR (AKA bit test and reset). Depends on the previous instruction.
+					// Or BLSR (AKA Reset Lowest Set Bit). No dependency but require BMI1 (basically a recent CPU)
+					p ^= 1 << j;
```

Comment:
I suppose I wonder whether this should actually be `p ^= 1U << j;` (I think 1 << 31 is undefined behaviour, but 1U << 31 is not).

As for merging -  You can go ahead, though I kinda want `std::` to be prefixed onto array and list ;)